File: 백준/Gold/14938. 서강그라운드/서강그라운드.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.StringTokenizer;


class Node implements Comparable<Node>{
	int n,d; // n for num, d for dist
	public Node(int n, int d) {
		this.n = n;
		this.d = d;
	}
	@Override
	public int compareTo(Node o) {
		return d - o.d;
	}
	
}

public class Main {
	static int n, range;
	static int[] value;
	static List<Node>[] map;
	
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		range = Integer.parseInt(st.nextToken());
		int route = Integer.parseInt(st.nextToken());
		value = new int[n + 1];
		st = new StringTokenizer(br.readLine());
		for (int i = 1; i <= n; i++) value[i] = Integer.parseInt(st.nextToken());
		map = new ArrayList[n+1];
		for (int i = 0; i <= n; i++) map[i] = new ArrayList<Node>();
		while (route --> 0) {
			st = new StringTokenizer(br.readLine());
			int start = Integer.parseInt(st.nextToken());
			int end = Integer.parseInt(st.nextToken());
			int dist = Integer.parseInt(st.nextToken());
			map[start].add(new Node(end,dist));
			map[end].add(new Node(start,dist));
		}
		int ans = 0;
		for (int i = 1; i <= n; i++) ans = Math.max(ans, dijikstra(i));
		System.out.println(ans);
	}
	
	static int dijikstra(int start) {
		int[] dist = new int[n+1];
		Arrays.fill(dist, -1);
		dist[start] = 0;
		Queue<Node> queue = new PriorityQueue<>();
		queue.add(new Node(start, 0));
		while (!queue.isEmpty()) {
			Node cur = queue.poll();
			if (dist[cur.n] == -1 || dist[cur.n] < cur.d) continue;
			for (Node next : map[cur.n]) {
				if (dist[next.n] == -1 || dist[next.n] > dist[cur.n] + next.d) {
					dist[next.n] = dist[cur.n] + next.d;
					queue.offer(new Node(next.n, dist[next.n]));
				}
			}
		}
		int cnt = 0;
		for (int i = 1; i <= n; i++) {
			if (dist[i] != -1 && dist[i] <= range) cnt += value[i];
		}
		return cnt;
	}
}

File: 백준/Gold/14938. 서강그라운드/서강그라운드.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const [n, m, r] = input[0].split(" ").map(Number);
const value = input[1].split(" ").map(Number);
const map = Array.from({ length: n + 1 }, () => []);

for (let i = 2; i <= r + 1; i++) {
    let [start, end, len] = input[i].split(" ").map(Number);
    map[start].push({ n: end, d: len }); 
    map[end].push({ n: start, d: len }); 
}

function dijkstra(start, n, range, value, map) {
    const dist = new Array(n + 1).fill(-1);
    dist[start] = 0;
    
    const queue = [{ n: start, d: 0 }];
    
    while (queue.length > 0) {
        queue.sort((a, b) => a.d - b.d);
        const cur = queue.shift();
        
        if (dist[cur.n] === -1 || dist[cur.n] < cur.d) continue;
        
        for (const next of map[cur.n]) {
            if (dist[next.n] === -1 || dist[next.n] > dist[cur.n] + next.d) {
                dist[next.n] = dist[cur.n] + next.d;
                queue.push({ n: next.n, d: dist[next.n] });
            }
        }
    }

    return Array.from({ length: n }, (_, i) => i + 1).reduce(
        (sum, i) => (dist[i] !== -1 && dist[i] <= range ? sum + value[i-1] : sum),  // Changed: value[i-1] since value array is 0-based
        0
    );
}

function findMaxValue(n, range, value, map) {
    return Math.max(...Array.from({ length: n }, (_, i) => i + 1)
        .map(i => dijkstra(i, n, range, value, map)));
}

console.log(findMaxValue(n, m, value, map));